### ISSUE TITLE
In RegexFeaturizer the regex_string seems not suitable for chinese?

### ISSUE THREAD

**Issue by journey1986** [RasaHQ/rasa#4069]
**Rasa version**:
"1.1.2"
**Rasa X version** (if used & relevant):

**Python version**:
3.7.3
**Operating system** (windows, osx, ...):
windows
**Issue**:
In RegexFeaturizer the regex_string seems not suitable for chinese?

In the RegexFeaturizer

```python
 # regex matching elements with word boundaries on either side
        regex_string = "(?i)(\\b" + "\\b|\\b".join(elements_sanitized) + "\\b)"
        return regex_string
```

It seems the above code not suitable for chinese.
```python
# regex matching elements with word boundaries on either side
        regex_string = "(?i)(\\b" + "\\b|\\b".join(elements_sanitized) + "\\b)"
        if self.language == 'zh':
            regex_string = "(?i)(" + "|".join(elements_sanitized) + ")"
        return regex_string
```
test code:
```python
import re

elements_sanitized = ["南京", "北京", "上海"]
text = "买一张从南京到上海的火车票"
regex_string1 = "(?i)(\\b" + "\\b|\\b".join(elements_sanitized) + "\\b)"
print(regex_string1)
matches1 = re.finditer(regex_string1, text)
matches1 = list(matches1)
print("matches1 len is:%d" % len(matches1))

regex_string2 = "(?i)(" + "|".join(elements_sanitized) + ")"
print(regex_string2)
matches2 = re.finditer(regex_string2, text)
matches2 = list(matches2)
print("matches2 len is:%d" % len(matches2))
```

**Error (including full traceback)**:
```

```

**Command or request that led to error**:
```

```

**Content of configuration file (config.yml)** (if relevant):
```yml

```

**Content of domain file (domain.yml)** (if relevant):
```yml

```

**Comment by msamogh** [comment]
Thanks for raising this issue, @ricwo will get back to you about it soon.

**Comment by ricwo** [comment]
Hi @journey1986, I don't know if I understand you correctly, but it seems that for Chinese characters word boundaries are not defined by spaces? Is this always true?

**Comment by journey1986** [comment]
Hi @ricwo , yes, you are right.

**Comment by ricwo** [comment]
@journey1986 I believe you should be able to use the `JiebaTokenizer` (see [here](https://rasa.com/docs/rasa/nlu/components/#jiebatokenizer)). You'll have to make sure to define a pipeline with `- name: "JiebaTokenizer"` followed by `- name: "RegexFeaturizer"`. Let me know if that works!

**Comment by journey1986** [comment]
@ricwo   I used JiebaTokenizer and it is also not working.
Below is my config.yml

```yml
# Configuration for Rasa NLU.
# https://rasa.com/docs/rasa/nlu/components/
language: "zh"
pipeline:
- name: "JiebaTokenizer"
  dictionary_path: "jieba_dict"
- name: "RegexFeaturizer"
- name: "CRFEntityExtractor"
- name: "EntitySynonymMapper"
- name: "CountVectorsFeaturizer"
  token_pattern: (?u)\b\w+\b
- name: "EmbeddingIntentClassifier"
  num_neg: 1

# Configuration for Rasa Core.
# https://rasa.com/docs/rasa/core/policies/
policies:
  - name: MemoizationPolicy
  - name: KerasPolicy
  - name: MappingPolicy
  - name: "FallbackPolicy"
    nlu_threshold: 0.3
    core_threshold: 0.3
    fallback_action_name: 'action_default_fallback'
```

**Comment by journey1986** [comment]
@ricwo  I have checked RegexFeaturizer features_for_patterns method source code:
```python
    def features_for_patterns(self, message):
        """Checks which known patterns match the message.

        Given a sentence, returns a vector of {1,0} values indicating which
        regexes did match. Furthermore, if the
        message is tokenized, the function will mark all tokens with a dict
        relating the name of the regex to whether it was matched."""

        found_patterns = []
        for exp in self.known_patterns:
            matches = re.finditer(exp["pattern"], message.text)
            matches = list(matches)
            found_patterns.append(False)
            for token_index, t in enumerate(message.get("tokens", [])):
                patterns = t.get("pattern", default={})
                patterns[exp["name"]] = False

                for match in matches:
                    if t.offset < match.end() and t.end > match.start():
                        patterns[exp["name"]] = True
                        found_patterns[-1] = True

                t.set("pattern", patterns)
```
you can see that it will first try to use regex_string to match the message.text.
if we use **regex_string = "(?i)(\\b" + "\\b|\\b".join(elements_sanitized) + "\\b)"**, then will not match the chinese sentence.

And then below code will not run:
```python
 for match in matches:
                    if t.offset < match.end() and t.end > match.start():
                        patterns[exp["name"]] = True
                        found_patterns[-1] = True
```

**Comment by ricwo** [comment]
Thanks! I've just tried to tokenize your sentence using jieba _without_ a custom dictionary. Here's what I get:
![image](https://user-images.githubusercontent.com/20581300/61847807-4bc01c00-aeac-11e9-895f-a98d491408a4.png)

It seems that both character pairs were tokenized correctly, so these should be correctly picked up by your `RegexFeaturizer`. Would you mind trying again without a custom dictionary, and parsing the exact same sentence you mentioned earlier, i.e. `"买一张从南京到上海的火车票"`?

**Comment by journey1986** [comment]
Hi @ricwo, jieba  is ok.But I means that,for example:
```python
exp["pattern"]="(?i)(\b南京\b|\b上海\b)"
message.text="买一张从南京到上海的火车票"?
matches = re.finditer(exp["pattern"], message.text)
```
Then matches is emtpy actually。

**Comment by ricwo** [comment]
Okay, thanks for clarifying that. It seems to me that instead of introducing the language config flag, wouldn't the simpler solution be to replace `message.text` in the line
```python
matches = re.finditer(exp["pattern"], message.text)
```
with the whitespace-joined message tokens, i.e.
```python
matches = re.finditer(exp["pattern"], " ".join(message.tokens))
```
Would you mind checking if this approach works for you, with of course the jieba tokeniser in the pipeline?

**Comment by journey1986** [comment]
@ricwo  I have checked the solution, it is right. But there is a mistake, we can not use message.tokens.
The right code is as below:
```python
            joined_tokens = " ".join(t.text for t in message.get("tokens", []))
            matches = re.finditer(exp["pattern"], joined_tokens)
```
Thank you for your reply :)

**Comment by xinghalo** [comment]
```
      regex_featurizer.py 

        # sanitize the regex, escape special characters
        elements_sanitized = [re.escape(e) for e in elements_to_regex]

        # regex matching elements with word boundaries on either side
        regex_string = "(?i)(\\b" + "\\b|\\b".join(elements_sanitized) + "\\b)"

        # add this can fix bug
        if self.language == 'zh':
            regex_string = "(?i)(" + "|".join(elements_to_regex) + ")"
        return regex_string
```

**Comment by journey1986** [comment]
@xinghalo  是的，我一开始也是用同样的方法规避的。 不过 @ricwo  提供了另一个不错的解决方案。

**Comment by ricwo** [comment]
@xinghalo @journey1986 would you mind adding some explanation to your two last comments?

**Comment by journey1986** [comment]
@ricwo   
translation：
@xinghalo Yes, I used the same method to circumvent it at first. However, @ricwo offers another great solution.

**Comment by ricwo** [comment]
@journey1986 thanks!

@erohmensing could we implement something like https://github.com/RasaHQ/rasa/issues/4069#issuecomment-516536896 into `RegexInterpreter` to avoid these things? In theory that should make it work with any language for which we have a tokeniser right?

**Comment by erohmensing** [comment]
@ricwo can you tl;dr how this is different from what we currently do? (I did read, but still don't quite understand)

**Comment by ricwo** [comment]
@erohmensing right now, we're matching on 

```python
matches = re.finditer(exp["pattern"], message.text)
```
and the question is can we simply switch to 
```python
joined_tokens = " ".join(t.text for t in message.tokens)
matches = re.finditer(exp["pattern"], joined_tokens)
```

this shouldn't make a difference for whitespace-tokenised sentences, but would mean Jieba-tokenised regex matching works

**Comment by erohmensing** [comment]
Sounds good to me! @journey1986 would you like to contribute a PR to include it?

**Comment by journey1986** [comment]
@erohmensing   I am sorry . I can not commit code to github from my company computer.

**Comment by erohmensing** [comment]
No worries, just wanted to offer the opportunity to contribute. Happy to take it on.

**Comment by wochinge** [comment]
@erohmensing That's not actually `in progress` right?

**Comment by erohmensing** [comment]
@wochinge mm i did actually start working on it, seemed to fix but there was also weird behavior i wanted to ask rick about. sorry i let it get stale! will tackle this week

**Comment by 8key** [comment]
Rasa takes regex definition from two places
1. e.g., 
```
## regex:loc
- 北京
- 南京
- 东京
```
2. in lookup table file
 e.g.,  https://github.com/RasaHQ/rasa/blob/f41326d21763d41ff92b8c469e434ab90d071a4d/data/test/lookup_tables/plates.txt

If you configure regex using the first method, there will be no issue (for Chinese), 
```
self.known_patterns = 
[{'name': 'hospital_specialist_name', 'pattern': '东京'}, {'name': 'hospital_specialist_name', 'pattern': '西京'}
```

if you configure regex using a lookup table, it will be issue (for Chinese, due to https://github.com/RasaHQ/rasa/blob/edd7750aba4cfb0555147223be7bc7dc1600be93/rasa/nlu/utils/pattern_utils.py#L54
),
```
self.known_patterns = 
[{'name': 'hospital_specialist_name', 'pattern': "(?i)(\b\南\京\b|\b\北\京\b|\b\东\京\b)"}]
``` 
It essentially combines the regex list defined in the file into one regex and that cause problems for any language that doesn't use white space for separation. 

A quick fix is just to disable this combination, modify the last two lines
```
def _generate_lookup_regex(lookup_table: Dict[Text, Union[Text, List[Text]]]) -> Text:
    """Creates a regex pattern from the given lookup table.
    The lookup table is either a file or a list of entries.
    Args:
        lookup_table: The lookup table.
    Returns:
        The regex pattern.
    """
    lookup_elements = lookup_table["elements"]

    # if it's a list, it should be the elements directly
    if isinstance(lookup_elements, list):
        elements_to_regex = lookup_elements
    # otherwise it's a file path.
    else:
        elements_to_regex = _read_lookup_table_file(lookup_elements)

    # sanitize the regex, escape special characters
    elements_sanitized = [re.escape(e) for e in elements_to_regex]
    lookup_regex_list = [{"name": table["name"], "pattern": i} for i in elements_sanitized]
    return lookup_regex_list
```

This will handled the issue.

**Comment by wochinge** [comment]
Mhm, I'm wondering what the performance difference is between one large regex and a list of small regexes 🤔

**Comment by 8key** [comment]
> Mhm, I'm wondering what the performance difference is between one large regex and a list of small regexes 🤔

Although they are essentially fast, preliminary test showed combined regex is faster. 
```
"""test 1m times comparing two regex schema, one is loop over small regex, one is combine regex in Or condition, take average value"""
import re, time

text = "买一张从南京到上海的火车票12345##, my phone number is 123-123-1234, email is abc@wee"

time_loop = []
time_combined = []
for i in range(1000000):
	b_time = time.time()
	for pattern in ['南京', '东京', '上海', "[0-9]{5}", "[*@&#]", "\\b(\d{1,3})?[-. (]*(\d{3})[-. )]*(\d{3})[-. ]*(\d{4})(?: *x(\d+))?\\b", "\w+@\w+"]:
		matches = re.finditer(pattern, text)
	time_loop.append(time.time() - b_time)

	b_time = time.time()
	matches = re.finditer("(?i)(南京|东京|上海|[0-9]{5})|[*@&#]|\\b(\d{1,3})?[-. (]*(\d{3})[-. )]*(\d{3})[-. ]*(\d{4})(?: *x(\d+))?\\b|\w+@\w+", text)
	time_combined.append(time.time() - b_time)
print('loop avg time', sum(time_loop)/len(time_loop))
print('combined avg time', sum(time_combined)/len(time_combined))
```
Result:
```
loop avg time 6.646946907043457e-06
combined avg time 1.062933921813965e-06
```

On the other hand, just noticed that if user defines regex in the intent file, namely, the first way, rasa wouldn't add word boundary \\b around, hence it worked for char-based languages. It is user's choice to surround with or without. However, in the second way, where user use look up tables, rasa would surround each regex with \\b, and in turn, caused the problem. Perhaps we need to remove this process in Rasa and let user determine in the look up table instead? In any way, rasa processes regex differently from two sources, and it can be problematic.

**Comment by koernerfelicia** [comment]
Hi, this issue was actually corrected in version 2.2.0 by [this PR](https://github.com/RasaHQ/rasa/pull/7422). You can now use the flag `use_word_boundaries` for `RegexFeaturizer` and `RegexEntityExtractor`.
